File: src/lib/mac-health.ts
import { prisma } from "@/lib/prisma";
import { Prisma } from "@prisma/client";
import { isUsageReadAuthorized } from "@/lib/ingest-auth";
import type { NextRequest } from "next/server";

export interface MacProcessStatus {
  name: string;
  status: "running" | "stopped" | "degraded" | "not_enabled" | "disabled" | "ok" | string;
  pid?: number | null;
  cpu?: number | null;
  memory?: number | null;
}

export interface MacHostTelemetry {
  hostname: string;
  username?: string;
  tailscaleHostname?: string;
  osVersion?: string;
  chipName?: string;
  arch?: string;
  cpuUsagePct: number;
  memoryUsagePct: number;
  diskUsagePct: number;
  uptimeSeconds: number;
  processes?: Record<string, "running" | "stopped" | "degraded" | "not_enabled" | "disabled" | string>;
  agentProcesses?: Record<string, "running" | "idle" | "stopped" | string>;
  pm2Processes?: Array<{ name: string; status: string; pid?: number | null; cpu?: number; memory?: number }>;
  launchdProcesses?: Array<{ name: string; status: string; pid?: number | null }>;
  lastHeartbeatAt: string;
}

export interface MacHealthResponse {
  ok: boolean;
  status: "online" | "degraded" | "offline";
  lastHeartbeatAt: string | null;
  secondsSinceHeartbeat: number | null;
  mac: MacHostTelemetry | null;
}

const ONLINE_THRESHOLD_SECONDS = 180; // 3 minutes

export async function recordMacHeartbeat(data: Partial<MacHostTelemetry>): Promise<MacHostTelemetry> {
  const now = new Date();
  const telemetry: MacHostTelemetry = {
    hostname: data.hostname || "jays.services",
    username: data.username || "jay",
    tailscaleHostname: data.tailscaleHostname || "macbook.boa-roygbiv.ts.net",
    osVersion: data.osVersion || "macOS",
    chipName: data.chipName || "Apple M5",
    arch: data.arch || "arm64",
    cpuUsagePct: Math.min(100, Math.max(0, Number(data.cpuUsagePct) || 0)),
    memoryUsagePct: Math.min(100, Math.max(0, Number(data.memoryUsagePct) || 0)),
    diskUsagePct: Math.min(100, Math.max(0, Number(data.diskUsagePct) || 0)),
    uptimeSeconds: Math.max(0, Math.floor(Number(data.uptimeSeconds) || 0)),
    processes: data.processes || {},
    agentProcesses: data.agentProcesses || {},
    pm2Processes: data.pm2Processes || [],
    launchdProcesses: data.launchdProcesses || [],
    lastHeartbeatAt: now.toISOString(),
  };

  const idempotencyKey = `mac-heartbeat-${telemetry.hostname}-${now.valueOf()}`;

  await prisma.externalUsageEvent.create({
    data: {
      idempotencyKey,
      sourceApp: "mac-host",
      provider: "local-mac",
      service: "mac-monitoring",
      metricType: "mac_heartbeat",
      billingMode: "estimated",
      confidence: "estimated",
      occurredAt: now,
      metadata: telemetry as unknown as Prisma.JsonObject,
    },
  });

  return telemetry;
}

export async function getLatestMacHealth(): Promise<MacHealthResponse> {
  const latestEvent = await prisma.externalUsageEvent.findFirst({
    where: {
      sourceApp: "mac-host",
      metricType: "mac_heartbeat",
    },
    orderBy: { occurredAt: "desc" },
    select: {
      occurredAt: true,
      metadata: true,
    },
  });

  if (!latestEvent || !latestEvent.metadata) {
    return {
      ok: false,
      status: "offline",
      lastHeartbeatAt: null,
      secondsSinceHeartbeat: null,
      mac: null,
    };
  }

  const metadata = latestEvent.metadata as unknown as MacHostTelemetry;
  const lastHeartbeatAt = new Date(metadata.lastHeartbeatAt || latestEvent.occurredAt);
  const secondsSinceHeartbeat = Math.floor((Date.now() - lastHeartbeatAt.getTime()) / 1000);

  let status: "online" | "degraded" | "offline" = "online";
  if (secondsSinceHeartbeat > ONLINE_THRESHOLD_SECONDS) {
    status = "offline";
  } else if (metadata.cpuUsagePct > 90 || metadata.memoryUsagePct > 90 || metadata.diskUsagePct > 95) {
    status = "degraded";
  }

  return {
    ok: status !== "offline",
    status,
    lastHeartbeatAt: lastHeartbeatAt.toISOString(),
    secondsSinceHeartbeat,
    mac: metadata,
  };
}

File: ios/UsageMonitor/UsageMonitorKit/Sources/Models/MacHealth.swift
import Foundation

/// `GET /api/health/mac` — latest Mac heartbeat (read token or dashboard session).
public struct MacHealthResponse: Codable, Hashable, Sendable {
    public var ok: Bool
    public var status: String
    public var lastHeartbeatAt: String?
    public var secondsSinceHeartbeat: Int?
    public var mac: MacHostTelemetry?

    public init(
        ok: Bool,
        status: String,
        lastHeartbeatAt: String? = nil,
        secondsSinceHeartbeat: Int? = nil,
        mac: MacHostTelemetry? = nil
    ) {
        self.ok = ok
        self.status = status
        self.lastHeartbeatAt = lastHeartbeatAt
        self.secondsSinceHeartbeat = secondsSinceHeartbeat
        self.mac = mac
    }

    public static let sample = MacHealthResponse(
        ok: true,
        status: "ok",
        lastHeartbeatAt: "2026-08-24T05:00:00.000Z",
        secondsSinceHeartbeat: 42,
        mac: MacHostTelemetry(
            hostname: "jays.services",
            username: "jay",
            tailscaleHostname: "macbook.boa-roygbiv.ts.net",
            osVersion: "macOS 15.6.1",
            chipName: "Apple M5",
            arch: "arm64",
            cpuUsagePct: 14.2,
            memoryUsagePct: 59.7,
            diskUsagePct: 87.0,
            uptimeSeconds: 86400 * 5 + 3600 * 4,
            processes: [
                "agent-sync": "running",
                "docker": "not_enabled",
                "litestream": "not_enabled",
                "ollama": "not_enabled"
            ],
            agentProcesses: [
                "claude": "running",
                "cursor": "running",
                "grok": "running",
                "codex": "running",
                "antigravity": "running",
                "copilot": "running"
            ],
            pm2Processes: [
                MacPm2Process(name: "agent-sync-live", status: "online", pid: 14205, cpu: 0.1, memory: 45000000),
                MacPm2Process(name: "fleet-heartbeat", status: "online", pid: 14206, cpu: 0.0, memory: 32000000)
            ],
            launchdProcesses: [
                MacLaunchdProcess(name: "com.jays.agent-sync-poll", status: "running", pid: 1205),
                MacLaunchdProcess(name: "com.jays.mac-watchdog", status: "running", pid: 1206)
            ],
            lastHeartbeatAt: "2026-08-24T05:00:00.000Z"
        )
    )
}

public struct MacPm2Process: Codable, Hashable, Sendable {
    public var name: String
    public var status: String
    public var pid: Int?
    public var cpu: Double?
    public var memory: Int?

    public init(name: String, status: String, pid: Int? = nil, cpu: Double? = nil, memory: Int? = nil) {
        self.name = name
        self.status = status
        self.pid = pid
        self.cpu = cpu
        self.memory = memory
    }
}

public struct MacLaunchdProcess: Codable, Hashable, Sendable {
    public var name: String
    public var status: String
    public var pid: Int?

    public init(name: String, status: String, pid: Int? = nil) {
        self.name = name
        self.status = status
        self.pid = pid
    }
}

public struct MacHostTelemetry: Codable, Hashable, Sendable {
    public var hostname: String
    public var username: String?
    public var tailscaleHostname: String?
    public var osVersion: String?
    public var chipName: String?
    public var arch: String?
    public var cpuUsagePct: Double
    public var memoryUsagePct: Double
    public var diskUsagePct: Double
    public var uptimeSeconds: Int
    public var processes: [String: String]?
    public var agentProcesses: [String: String]?
    public var pm2Processes: [MacPm2Process]?
    public var launchdProcesses: [MacLaunchdProcess]?
    public var lastHeartbeatAt: String?

    public init(
        hostname: String,
        username: String? = nil,
        tailscaleHostname: String? = nil,
        osVersion: String? = nil,
        chipName: String? = nil,
        arch: String? = nil,
        cpuUsagePct: Double,
        memoryUsagePct: Double,
        diskUsagePct: Double,
        uptimeSeconds: Int,
        processes: [String: String]? = nil,
        agentProcesses: [String: String]? = nil,
        pm2Processes: [MacPm2Process]? = nil,
        launchdProcesses: [MacLaunchdProcess]? = nil,
        lastHeartbeatAt: String? = nil
    ) {
        self.hostname = hostname
        self.username = username
        self.tailscaleHostname = tailscaleHostname
        self.osVersion = osVersion
        self.chipName = chipName
        self.arch = arch
        self.cpuUsagePct = cpuUsagePct
        self.memoryUsagePct = memoryUsagePct
        self.diskUsagePct = diskUsagePct
        self.uptimeSeconds = uptimeSeconds
        self.processes = processes
        self.agentProcesses = agentProcesses
        self.pm2Processes = pm2Processes
        self.launchdProcesses = launchdProcesses
        self.lastHeartbeatAt = lastHeartbeatAt
    }

    /// Sorted process rows so the UI does not shuffle on every refresh.
    public var processRows: [(name: String, status: String)] {
        (processes ?? [:]).map { ($0.key, $0.value) }.sorted { $0.name < $1.name }
    }
}
